src/miniscript/mod.rs

```diff
@@ -433,6 +467,7 @@ mod tests {
     use std::sync::Arc;
 
     type Segwitv0Script = Miniscript<bitcoin::PublicKey, Segwitv0>;
+    type Tapscript = Miniscript<bitcoin::schnorr::PublicKey, Tap>;
```

Comment:
I think we need somehow prevent users from using `Miniscript<bitcoin::schnorr::PublicKey, SegwitV0>`-like types...

Non-clean route will be to provide a set of type aliases for the supported combinations, but this still will not prevent users from constructing incorrect onces.

The cleaner way is much more complex though: something like creating more traits, one per context, for public keys; and moving first `Miniscript` generic parameter to be a generic parameter of the context, restricted by the public key trait...

Reply:
Agreed, this can be a problem and can be confusing. But even if users do that, they cannot shoot themselves in foot. All the current APIs would work correctly. The encoding API would give a 33-byte public key when the user calls `ms.encode()` and decoding would be looking for 33 byte pubkey keys and would error on `schnorr::PublicKey`. All the APIs would output correctly assuming the context is `Segwitv0`. 

All `MiniscriptKey` is "provide me some basic bounds(FromStr/Clone/etc) way to convert to `bitoin::PublicKey`". Users can free to any `impl MiniscriptKey` the user wants. If the wants to only use only a enum ```
enum AllowedKeys
{
AliceKey,
BobKey,
CarolKey,
}
```
and provide a `impl MiniscriptKey` for it, the library would just work. Similarly, for x-only keys, the output would not be wrong, we are not breaking any safety rules, just that the encoded output might be in `SegwitV0` context with 33-byte keys which might be confusing. 

If the user then goes on to the encoded bytes into a tapscript script, I think it is a user error. Only advanced users should be dealing with the `Miniscript` object, and users that deal with descriptors directly should would very clearly know that they are dealing with `tr` descriptor or `wsh` one.

Reply:
I think we should need API users ability to restrict that their custom `impl MiniscriptKey` keys like `AllowedKeys` should be restricted only for Taproot or other context. For instance, this may be required when they are derived under certain BIP32 path requiring Taproot or SegWit usage. And the second (more complex) way I proposed should work in this case as well

Reply:
How about the following which is less invasive, but yet another hack. 
- We add a method to MiniscriptKey `is_full_key` with a default impl that returns True.
- We implement `is_full_key` as false for `schnorr::PublicKey`
- Inside the global checks, we add check for keys in PkK, MultiA.

This is not clean because this does not disable this at type level, nor it allows downstream users to restrict implementation based on context. But I think this can fall within the scope of this PR if we think it is the right thing to do.

However, I am still not completely convinced that `Miniscript<bitcoin::schnorr::PublicKey, Tap>` should be prevented. In my head, MiniscriptKey is still "provide me some basic bounds(FromStr/Clone/etc) way to convert to bitoin::PublicKey" which is still fits for `schnorr::PublicKey`. cc @apoelstra

Reply:
Maybe we can add an associate type to ScriptContext which is "ConesnusKey"?

And then you can have methods that are only defined when `Miniscript<K, C> /*...*/  where K : IsEqual<C::ConsensusKey>`, or perhaps `Miniscript<K, C> /*...*/  where C::ConsensusKey : From<K>`, or `Miniscript<K, C> /*...*/  where C::ConsensusKey : From<(K, Option<KeyDerivationStruct>)>`. There's obviously some design flexibility here, but the end outcome is that ScriptContexts know what kind of key they need for consensus purposes.

Reply:
Sounds nice. Let me fiddle around with this.

Reply:
Thanks for the suggestion. Ended up adding an associated `Key` type to the `ScriptContext` trait.

Reply:
looks good